[WFCORE-1579] Perform check when changing 'path' attribute of deploym…

TomasHofman opened the pull request:
…ent-scanner instance

https://issues.jboss.org/browse/WFCORE-1579
https://issues.jboss.org/browse/JBEAP-4874

Changes to the deployment scanner so that if deployment dir is not accessible:
* bootTimeScan() is skipped and error is logged (before the server failed to start)
* establishDeployedContentList() is not performed during startScanner(), instead it will be performed during first scan when deployment dir is accessible

TomasHofman commented:
Failing tests in the Wildfly testsuite need to be modified to reflect changes in this PR.

bstansberry commented on deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java:
```diff
@@ -780,6 +804,44 @@ private boolean handleCompositeResult(ModelNode resultNode, StringBuilder failur
+        else if (!directory.canWrite()) {
+            if (deploymentDirAccessible) {
+                deploymentDirAccessible = false;
+                ROOT_LOGGER.directoryIsNotWritable(deploymentDir.getAbsolutePath());
```
There needs to be a parameter or something to disable the logging from this method if the information has already been logged. Otherwise it looks like the same logging will happen 3 times at boot (bootTimeScan, startScanner, scan) and then periodically thereafter every time the scanner runs.

bstansberry replied:
Never mind; I'm not paying attention; that's what deploymentDirAccessible does!

bstansberry commented on deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java:
```diff
@@ -567,6 +576,21 @@ private boolean scan(boolean oneOffScan, final DeploymentOperations deploymentOp
+            // if deployed content list was not established during scanner start (due to inaccessible deployment dir),
+            // do it now
+            if (!deployedContentEstablished) {
+                establishDeployedContentList(deploymentDir, deploymentOperations);
```
@ehsavoie Can you review this PR please?

This bit here is the only concern I have. Not that it's necessarily wrong. Just that this is the one change where I think careful review is needed as it could introduce subtle bugs. And, if it's ok to do this here, then it seems unnecessary to do an establishDeployedContentList in startScanner() and L399-406 can just be dropped.

bstansberry commented:
Thanks for this, Tomas!

bstansberry commented:
retest this please

bstansberry commented:
https://github.com/wildfly/wildfly/pull/9071 is a PR to remove that test in full; #1701 is to add it to core. So this PR should be updated to update the test that will be in core.

TomasHofman commented:
I updated FileSystemDeploymentService according to Brian's suggestion (left out establishDeployedContentList() call from FileSystemDeploymentService#startScanner()) - it required few more changes in FileSystemDeploymentServiceUnitTestCase - and fixed the test case moved from full.

TomasHofman commented:
retest this please

TomasHofman commented:
retest this please
